### PR TITLE
Qualified defs

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Allow qualified names in LHS of type signatures and method definitions.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 1]
```diff
@@ -807,7 +807,7 @@ yield switch (ast) {
         case IR.Expression e -> e;
       };
       case Tree.TypeSignature sig -> {
-        var methodName = buildName(sig, sig.getVariable());
+        var methodName = buildName(sig.getVariable());
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:854]: This is OK, I guess. Now the location information can be extracted from a `Token`, so we don't need the two argument variant anymore.

File: engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java [hunk 2]
```diff
@@ -431,6 +431,24 @@ public void testTypeSignature() throws Exception {
     """);
   }
 
+  @Test
+  @Ignore
```
Comment (JaroslavTulach) [inline engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java:509]: I see. Neither `resolve : ...` nor `id x = x` is supported. I'll work on that.

**Review by JaroslavTulach — APPROVED** [review]

File: lib/rust/parser/debug/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    let number_base_token = rust_to_meta[&token::variant::NumberBase::reflect().id];
+    let newline_token = rust_to_meta[&token::variant::Newline::reflect().id];
+    let text_start_token = rust_to_meta[&token::variant::TextStart::reflect().id];
+    let text_end_token = rust_to_meta[&token::variant::TextEnd::reflect().id];
+    let text_section_token = rust_to_meta[&token::variant::TextSection::reflect().id];
+    let text_escape_token = rust_to_meta[&token::variant::TextEscape::reflect().id];
+    let wildcard_token = rust_to_meta[&token::variant::Wildcard::reflect().id];
+    let autoscope_token = rust_to_meta[&token::variant::AutoScope::reflect().id];
+    // TODO: Implement `#[reflect(flag = "enso::concrete")]`, which just attaches user data to the
+    //  type info; then filter by flag here instead of hard-coding these simplifications.
```
Comment (wdanilo) [inline lib/rust/parser/debug/src/lib.rs]: todo lets in the code. If you want to leave it for another PR, create issue and link it  here please.

File: lib/rust/parser/debug/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    let close_symbol_token = rust_to_meta[&token::variant::CloseSymbol::reflect().id];
+    let number_token = rust_to_meta[&token::variant::Digits::reflect().id];
+    let number_base_token = rust_to_meta[&token::variant::NumberBase::reflect().id];
+    let newline_token = rust_to_meta[&token::variant::Newline::reflect().id];
+    let text_start_token = rust_to_meta[&token::variant::TextStart::reflect().id];
+    let text_end_token = rust_to_meta[&token::variant::TextEnd::reflect().id];
+    let text_section_token = rust_to_meta[&token::variant::TextSection::reflect().id];
+    let text_escape_token = rust_to_meta[&token::variant::TextEscape::reflect().id];
+    let wildcard_token = rust_to_meta[&token::variant::Wildcard::reflect().id];
+    let autoscope_token = rust_to_meta[&token::variant::AutoScope::reflect().id];
```
Comment (wdanilo) [inline lib/rust/parser/debug/src/lib.rs]: this looks like a repetition - can we automate it somehow?

File: lib/rust/parser/debug/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    to_s_expr.skip(wildcard_token);
+    to_s_expr.skip(autoscope_token);
+    to_s_expr.skip(text_start_token);
+    to_s_expr.skip(text_end_token);
+    to_s_expr.skip(open_symbol_token);
+    to_s_expr.skip(close_symbol_token);
+    tuplify(to_s_expr.value(ast_ty, &value))
+}
+
+/// Strip certain fields that should be excluded from output.
```
Comment (wdanilo) [inline lib/rust/parser/debug/src/lib.rs]: What do you mean by "should be excluded from output"? These docs do not explain what this function does.

File: lib/rust/parser/debug/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+/// Given an S-expression representation of a [`Token`] and the base address for `Code` `Cow`s,
+/// return the range of the input code the token references.
+fn token_code_range(token: &Value, base: usize) -> std::ops::Range<usize> {
+    let get_u32 =
+        |field| fields(token).find(|(name, _)| *name == field).unwrap().1.as_u64().unwrap() as u32;
+    let begin = get_u32(":codeReprBegin");
+    let len = get_u32(":codeReprLen");
+    let begin = (begin as u64) | (base as u64 & !0xFFFF_FFFF);
+    let begin = if begin < (base as u64) { begin + 0x1_0000_0000 } else { begin };
```
Comment (wdanilo) [inline lib/rust/parser/debug/src/lib.rs]: magic literals

**Review by wdanilo — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 1 above)
Comment (kazcw) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:854]: I had to change it because the `variable` field is a `Tree` now, since it may be an `Ident` or `OprApp`s

File: engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java (hunk 2 above)
Comment (JaroslavTulach) [reply engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java:509]: I have the implementation. It works, but a single issue - I need to check if an identifier is an operator or not. The code is here https://github.com/enso-org/enso/blob/3e03e7db3a1237e14bf20dcc207351c45bad81e1/engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java#L1826 - is there any recommended way to find out a `Token` is an operator?

I am getting `Token.Ident` even for `==` with `toString()` value `Ident["", "==", false, 0]`

Comment (JaroslavTulach) [reply engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java:509]: For now recognized manually: https://github.com/enso-org/enso/pull/3611/commits/51efba96174488d678b687fdc9850a77bb139d86

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 7]
```diff
@@ -685,31 +685,6 @@ yield switch (identifier) {
               yield new IR$Function$Lambda(args, body, getIdentifiedLocation(tree), true, meta(), diag());
             }
           }
-          case "=" -> {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:688]: Great, thanks!
